kubernetes/client_factory.go

```diff
@@ -253,3 +288,50 @@ func getTokenHash(authInfo *api.AuthInfo) string {
 	}
 	return string(h.Sum(nil))
 }
+
+// KialiSAClients returns all clients associated with the Kiali service account across clusters.
+func (cf *clientFactory) GetSAClients() map[string]ClientInterface {
+	cf.mutex.RLock()
+	defer cf.mutex.RUnlock()
+
+	for cluster := range cf.saClientEntries {
+		if err := cf.refreshClientIfTokenChanged(cluster); err != nil {
+			log.Errorf("Unable to refresh Kiali SA client for cluster: %s. Err: %s", cluster, err)
+		}
+	}
+
+	return cf.saClientEntries
+}
+
+// Check for kiali token changes and refresh the client when it does.
+func (cf *clientFactory) refreshClientIfTokenChanged(cluster string) error {
+	kialiSAToken, err := GetKialiToken()
+	if err != nil {
+		return err
+	}
+
+	if cf.saClientEntries[cluster].GetToken() != kialiSAToken {
```

Comment:
I think a `TODO` is missing somewhere around this func, no?

If I understand correctly, it tries to compare the token of a remote cluster client, with the Kiali SA token (which would be for the local cluster). That's always going to be a mismatch, no? So, it will always want to recreate the client.

(this is one of the reasons why I think `clientfactory` is not ready for MC yet... although I see you have a unit test for this, so it is confusing).

Reply:
Yeah at the very least it needs a `TODO`. It should not be comparing to the kiali token for each cluster. Rather it should be comparing the client's token with the token read from the filesystem for that cluster. The unit for the home cluster is valid but the one for multiple clients is a bit misleading since we won't want to test with kiali token but rather with the token for that cluster. I'll remove the multiple cluster test since that will change anyways once the client factory supports reading the secrets from the fs.

Reply:
Well looking at the test again it's actually testing two different methods. One test is for `GetSAHomeClusterClient` and the other test is for `GetSAClients`. In both cases we want to test that you get an updated client back in return but I'll update the test for `GetSAClients` so it's more clear.